core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -344,6 +344,9 @@ public int from() {
      * The number of search hits to return. Defaults to <tt>10</tt>.
      */
     public SearchSourceBuilder size(int size) {
+        if (size < 0) {
+            throw new IllegalArgumentException("[size] parameter cannot be negative");
+        }
```

Comment:
Okay, the set/unset distinction makes sense e.g. when using the builder on the client side, otherwise we would always render the `size` even if not set. What makes me not like this is the fact that we reject "-1" as illegal but still use it as a "legal" value meaning something internally. To me it feels a bit weird when you are just looking at this class.